torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp

```diff
@@ -0,0 +1,211 @@
+    {c10::kFloat, 1},
+    {c10::kByte, 2},
+    {c10::kChar, 3},
+    {c10::kShort, 5},
+    {c10::kInt, 6},
+    {c10::kLong, 7},
+    {c10::kBool, 9},
+    {c10::kHalf, 10},
+    {c10::kDouble, 11},
+};
+
+static int64_t ScalarTypeToONNXType(const c10::ScalarType& st) {
+  int64_t onnx_type = -1;
+  if (scalarTypeToONNXTypeMap.count(st) != 0) {
+    onnx_type = scalarTypeToONNXTypeMap.find(st)->second;
+  }
+  return onnx_type;
+}
+
+static const std::unordered_set<NodeKind> arithmeticOps = {
+  onnx::Add,
+  onnx::Sub,
+  onnx::Mul,
+  onnx::Div,
+  onnx::Gemm,
+  onnx::Pow,
+  onnx::Mod,
+};
+
+static bool IsArithmeticOp(const NodeKind& nkind) {
+  return arithmeticOps.find(nkind) != arithmeticOps.end();
+}
+
+static const std::unordered_set<NodeKind> comparisonOps = {
+  onnx::Greater,
+  onnx::Less,
+  onnx::Equal,
+};
+
+static bool IsComparisonOp(const NodeKind& nkind) {
+  return comparisonOps.find(nkind) != comparisonOps.end();
+}
+
+static ProfiledTensorTypePtr CreateProfiledTensorTypeWithScalarType(
+    const ProfiledTensorTypePtr& typePtr,
+    const c10::ScalarType& scalar_type) {
+  return ProfiledTensorType::create(
+      scalar_type,
+      typePtr->device(),
+      typePtr->sizes(),
+      typePtr->strides(),
+      typePtr->requiresGrad());
+}
+
+static bool IsImplicitCastSupported(const NodeKind& nodeKind) {
+  return (arithmeticOps.find(nodeKind) != arithmeticOps.end() ||
+      comparisonOps.find(nodeKind) != comparisonOps.end());
+}
+
+static c10::optional<c10::ScalarType> PromoteScalarTypes(const std::vector<c10::ScalarType>& types) {
+  if (types.empty()) {
+    return c10::nullopt;
+  }
+  auto st = types[0];
+  for (size_t i=1; i<types.size(); ++i) {
+    st = c10::promoteTypes(st, types[i]);
+  }
+  return st;
+}
+
+static c10::optional<c10::ScalarType> InferExpectedScalarType(const Node* n) {
+  std::vector<c10::ScalarType> typesFromTensors;
+  std::vector<c10::ScalarType> typesFromScalars;
+  std::for_each(n->inputs().begin(), n->inputs().end(), [&](const Value* input){
+    auto nkind = input->node()->kind();
+    if (nkind == onnx::Gather && input->node()->input(0)->node()->kind() == onnx::Shape) {
+      typesFromScalars.emplace_back(c10::kLong);
+    } else if (nkind == onnx::Constant) {
+      typesFromScalars.emplace_back(input->node()->t(attr::value).scalar_type());
+    } else if (auto scalar_type = ProfiledTensorType::create(input->type())->scalarType()) {
+      typesFromTensors.emplace_back(*scalar_type);
+    }
+  });
+
+  c10::optional<c10::ScalarType> st = c10::nullopt;
+  const c10::optional<c10::ScalarType> output_st = ProfiledTensorType::create(n->output()->type())->scalarType();
+
+  if (typesFromScalars.size() == n->inputs().size()) {
+    // If all inputs are scalars, infer scalar_type by calling c10::promoteTypes.
+    st = PromoteScalarTypes(typesFromScalars);
+  } else if (output_st && !IsComparisonOp(n->kind())) {
+    st = output_st;
+  } else if (!typesFromTensors.empty()) {
```

Comment:
@BowenBao could you please put a comment here explaining that this is supposed to capture the scenario when one input is a scalar and another is a tensor.

Reply:
thanks, added